# Recommendation Systems


![Title](img/title.jpg)

> ## Syed Owais Chishti
## Data Scientist.

### Hello every and welome to another tutorial. This week we are going to talk about how to increase sales/purchase of products.

## but you might be thinking how AI can help in this?

### But do not forget, Artificial Intelligence is Super Man.
![man](img/man.jpg)

### We can use a technique called "Recommendations System" to resolve this particular problem, and further there are many algorithms used as RS (will use RS as short form of Recommendation System through out tutorial). Like, 
* LightFM
* Apriori
* Eclat etc.

## So, why are we waiting? Let's simply dive into it.

## If you still do not have idea how RS works, simply try to understand the meaning of following sentence.

> **"people who bought also bought"**

** above line is the key idea of RS, it helps you in purchasing relevant goods and helps seller so that he can put relevant items at one place**

** Got it ? Ok let's take an example, few years ago it was found that two products were sold rapidly in evening timings,**
* Baby Diapers.
* Beer (beverage).

## It maked me shoked !!!
![shoc](img/shoc.jpg)

** But, the fact is, normally in evenings man went out to purchase baby diapers and due to tiredness they bought beer.**

## AI rocks.
![rocks](img/rocks.jpg)

## Types of RS

* Collaborative.
* Content Based.

** Let's discuss a bit about types.**

### Collaborative:

** It predicts your likeness on the basis of other people likenesses due to their purchases in past.**

### Content Based.

** It predicts your likeness on the basis of your likenesses due to your purchases in past.**

### All the big companies like Google (for search engine), Amazon (for products), Youtube (for videos), Facebook (for relevant friends) use this RS technology with both Collaborative and Content Based types to improve user activities.
![all](img/all.jpg)

### Let us now see how RS works by coding it with my love "PYTHON".

** We are going to use two types of Algorithms,**
* LightFM and
* Apriori.

** with movies and grocery items respectively.**

** I am using Anaconda Enviroment with Python 3.6.x., to download anaconda check out my prevoius tutorials.
But LightFM library is not available with Anaconda, you need to download its .whl file** [here](https://ci.appveyor.com/api/buildjobs/uj88ko5l53rmqdkt/artifacts/dist%2Flightfm-1.13-cp36-cp36m-win_amd64.whl).

** After you installed .whl file go to the directory where it is placed and press "Shift+Rightclick" and press "open command window here" (This is for 64-bit windows, for MAC open terminal in current directory) and enter**

> ** pip install lightfm-1.13-cp36-cp36m-win_amd64.whl**

In [1]:
## Apriori algorithm with grocery data set.

# first install necessary libraries.
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# read data set
dataset = pd.read_csv('Market_Basket_Optimisation.csv', header=None)

transactions = []
for i in range(0,7501):
    transactions.append([str(dataset.values[i,j]) for j in range(0,20)])

from apyori import apriori
rules = apriori(transactions, min_support=0.003, min_confidence=0.2, min_lift=3, min_lenght=2)# min_lenght is to set no. of relevant products

results = list(rules)

myResults = [list(x) for x in results]

for items in range(10):
    print(myResults[items])

[frozenset({'chicken', 'light cream'}), 0.004532728969470737, [OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]]
[frozenset({'escalope', 'mushroom cream sauce'}), 0.005732568990801226, [OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]]
[frozenset({'pasta', 'escalope'}), 0.005865884548726837, [OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]]
[frozenset({'honey', 'fromage blanc'}), 0.003332888948140248, [OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0.2450980392156863, lift=5.164270764485569)]]
[frozenset({'ground beef', 'herb & pepper'}), 0.015997866951073192, [OrderedStatistic(items_base=frozenset({'herb & pepper'}), items_add=frozenset({'g

> ** [frozenset({'chicken', 'light cream'}), 0.004532728969470737, [OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]] **

** In this line you can clearly see that people who bought chicken also bought ice cream (may they used to eat ice cream while waiting during order).**

**To understand other parts of answer let's understand few basic parts of apriori**

* Support. If we talk about movie recommendations (M), then, support(M) = # of watchlists containing(M) / # of user watchlists.
** Assume 10 people saw the movie M out of 100. support(M) = 10/100 = 10%**


* Confidence (from movie1 M1 to movie2 M2). confidence(M1-->M2) = # of watchlists of M1 and M2 / # of user watchlists of M1.
** Say 80 people saw movie M1 and 7 people saw movie M2 and M1 both. confidence(M1-->M2) = 7/80 = 17.5%**


* Lift(M1-->M2) = confidence(M1-->M2) / support(M2).
** Lift(M1-->M2) = 17.5 / 10 = 1.75**


** Lift gives the ratio of suggestions, the higher the ratio of lift the more suggestions to person for movie2 already seen movie1**.

> **rules = apriori(transactions, min_support=0.003, min_confidence=0.2, min_lift=3, min_lenght=2)**

** All three parameters of apriori can be tuned according to your requirements.**

## Now let's see LightFM Algorithm.

In [9]:
import numpy as np
from lightfm.datasets import fetch_movielens
from lightfm import LightFM

data = fetch_movielens(min_rating=4.0) # to avoid space issues using only those movies which are rated 4 or above.

print(repr(data['train']))

print(repr(data['test']))

model = LightFM(loss='warp')

model.fit(data['train'], epochs=30, num_threads=2)

def RS(model, data,user_ids):
    n_users, n_items = data['train'].shape

    for user_id in user_ids:
        known_positives = data['item_labels'][data['train'].tocsr() [user_id].indices]
        scores = model.predict(user_id, np.arange(n_items))
        top_items = data['item_labels'][np.argsort(-scores)]

        print("User %s" % user_id)
        print(" Watched before")
        for x in known_positives[:3]:
            print ("     %s" % x)
        print(" Recommended for user: ")
        for x in top_items[:3]:
            print("     %s" % x)
RS(model, data, [3,25,450])

<943x1682 sparse matrix of type '<class 'numpy.int32'>'
	with 49906 stored elements in COOrdinate format>
<943x1682 sparse matrix of type '<class 'numpy.int32'>'
	with 5469 stored elements in COOrdinate format>
User 3
 Watched before
     Seven (Se7en) (1995)
     Contact (1997)
     Starship Troopers (1997)
 Recommended for user: 
     Contact (1997)
     Scream (1996)
     G.I. Jane (1997)
User 25
 Watched before
     Dead Man Walking (1995)
     Star Wars (1977)
     Fargo (1996)
 Recommended for user: 
     Contact (1997)
     Fargo (1996)
     English Patient, The (1996)
User 450
 Watched before
     Contact (1997)
     George of the Jungle (1997)
     Event Horizon (1997)
 Recommended for user: 
     G.I. Jane (1997)
     Kiss the Girls (1997)
     Scream (1996)


### Both Apriori and LightFM are good algorithms to start, but LightFM is better due to it contains several data sets in it. Though both algorithms can give you kick to start using RS in your AI works.

** WARP above is "weighted Approximate-Rank Pairwise", It helps to predict on the basis of previous other users likes. It uses Gradient descent to iteratively find the weights which improve our predictions over time.**


## Topic for next week is **Decentralized Applications (Dapp)**.

## Hope you enjoyed this tutorial.


## Check my other tutorials as well, and also give your reviews on my research papers.

* [More Tutorials on GitHub](https://github.com/elacsoft)
* [Research Papers](https://github.com/elacsoft/Research-Papers)

## Syed Owais Chishti
### Data Scientist.

>* [MyWebsite](http://elacsoft.cf) 
* [MyTwitterAccount](https://twitter.com/robertw26984557)
* [MyFacebookAccount](https://www.facebook.com/owais.chishti.35)
* [MyGitHubAccount](https://github.com/elacsoft)
* [MyYoutubeChannel](https://www.youtube.com/channel/UCWha7dSnT_k7mBGJ0XFB0IQ)
* [elacsoft@gmail.com](elacsoft@gmail.com)